File: HL_scripts/splitChain_into_randomParts.py
#!/usr/bin/env python3
"""Replacement for splitChain_into_randomParts.pl

I don't write in Perl so rewriting the script in Python
is more time and resources efficient than just rewriting
the original script.
"""
import argparse
import os
import sys
import random

__author__ = "Bogdan Kirilenko, perl version: Nikolai Hecker"

DESCRIPTION = """Splits a chain file into n chain split files,
randomly picking chain ids.\nExample usage:\n
$0 -c hg38.speTri2.all.chain -n 100 -p splitChains_hg38_speTri2_\n\n";
Given very large chains this script might be somewhat heavy
on memory for storing millions of chainIDs.\n";
"""


def parse_args():
    app = argparse.ArgumentParser(description=DESCRIPTION)
    app.add_argument("--chain", "-c", help="Chain file to be split")
    app.add_argument("--prefix", "-p", help="Prefix for output files")
    app.add_argument("--nsplit",
                     "-n", 
                     help="Number desired files chain is split into")
    if len(sys.argv) < 4:
        app.print_help()
        sys.exit(0)
    args = app.parse_args()
    if any(x is None for x in (args.chain, args.prefix, args.nsplit)):
        # !! I recapitulated the perl script argument parset procedure
        # for compatibility, of course it would be better to explictly
        # set these arguments as mandatory
        app.print_help()
        sys.exit(0)
    return args


def split_in_n_lists(lst, n):
    """Split a list into n list of more or less equal length."""
    if n <= 0:  # must never happen
        sys.stderr.write("Error! Method split_in_n_lists called with n == 0\n")
        sys.exit(1)
    lst_len = len(lst)
    if n >= lst_len:
        # pigeonhole principle in work
        return [[x] for x in lst]
    ret = []  # list of lists
    sublist_len = lst_len / float(n)
    last = 0.0
    while last < len(lst):
        sublist = lst[int(last) : int(last + sublist_len)]
        ret.append(sublist)
        last += sublist_len
    return ret


def read_chain_ids(chain_file):
    """Just grab a list of chain IDs in the present file."""
    ret = []
    f = open(chain_file, 'r')
    for line in f:
        if not line.startswith('chain'):
            continue
        ld = line.rstrip().split(" ")
        ret.append(int(ld[-1]))
    f.close()
    return ret


def main():
    """Entry point."""
    args = parse_args()
    chain_ids = read_chain_ids(args.chain)


if __name__ == "__main__":
    main()

File: HL_scripts/splitChain_into_randomParts.pl
#!/usr/bin/env perl
#
#
#
# <Nikolai Tue Jan 30 17:41:09 CET 2018>
#
use strict;
use Carp;
use Getopt::Long;
use POSIX;
use List::Util qw/shuffle/;

### options
my $inf = "";
my $prefix = "";
my $n;
undef $n;

my $usage = "Usage:\n\t$0 [OPTIONS]\n";
$usage .= "\t-c|chain\t\tFILENAME\tChain file to be split\n";
$usage .= "\t-p|prefix\t\tSTRING\t\tPrefix for output files\n";
$usage .= "\t-n|nsplit\t\tINTEGER\t\tNumber desired files chain is split into\n\n";
$usage .= "Splits a chain file into n chain split files, randomly picking chain ids.\nExample usage:\n $0 -c hg38.speTri2.all.chain -n 100 -p splitChains_hg38_speTri2_\n\n";
$usage .= "Given very large chains this script might be somewhat heavy on memory for storing millions of chainIDs.\n";

GetOptions(
    "c|chain=s" =>  \$inf,
    "p|prefix=s" => \$prefix,
    "n|nsplit=i" => \$n
    );

# check mandatory options
if($inf eq "" || $prefix eq "" || !defined($n) )
{
    print $usage;
    croak "Too few or wrong arguments.";
}


### main
my @achainIds = (); # store chain IDs
my %h_idfh = (); # hash will assign chain IDs to file handle array positions
my @fhs = (); # file handles

my $fh;

# get chain IDs
open($fh, "<$inf") || croak "Can't open chain file '$inf'.";

while(<$fh>)
{
    my $line = $_;
    chomp($line);

    # store chain ID
    if($line =~ /^chain.*\s+(\d+)$/)
    {
	my $id = $1;
	push(@achainIds, $id);
    }
}
close($fh);

my $nids = @achainIds;
print "Found $nids chain IDs\n";

# shuffle ids
@achainIds = shuffle @achainIds;

# create file handle hash
my $pos = 0;

for(my $i=0; $i < @achainIds; $i++)
{
    $h_idfh{$achainIds[$i]} = $pos;
    
    # increment pos or set it back to 0
    $pos = ($pos + 1) < $n ? ($pos + 1) : 0;
}


### open all file handles
for(my $i=0; $i < $n; $i++)
{
    my $outf = $prefix.$i;
    my $tfh;
    open($tfh, ">$outf") || croak "Failed to open file handle for split file '$outf'";
     $fhs[$i] = \$tfh; # store file handle references
    
}

### process chain file and store write to assigned file handle
open($fh, "<$inf") || croak "Can't open chain file '$inf'.";

while(<$fh>)
{
    my $line = $_;
    chomp($line);

    # store chain ID
    if($line =~ /^chain.*\s+(\d+)$/)
    {
	# get file handle 
	my $id = $1;
	my $fhpos = $h_idfh{$id};
	my $refoutfh = $fhs[$fhpos];
	my $outfh = $$refoutfh;

	# write chain header
	print $outfh "$line\n";

	# get process lines and write output
	while( ($line = <$fh>) =~ /^\d+/)
	{
	    print $outfh "$line";
	}
	#print final empty line
	print $outfh "\n";
    }
}
close($fh);


### close all file handles
for(my $i=0; $i < $n; $i++)
{
    my $reftfh = $fhs[$i];
    my $tfh = $$reftfh;
    close($tfh);
}

print "Wrote output to $n files starting with '$prefix'.\n"
